### PR TITLE
Feature/geinfra 376

### PR DESCRIPTION
**Pull request by AltusBarry**
Updates to site deletion_method data and id validation.

Update to site model to handle create AND update, but raises normal python exceptions.

Added validator that get used in the controllers themselves.
Added a custom exception and handler, registered to the connexion app in `swagger_server/test/__init__.py` and `swagger_server/__main__.py`

Flake8 cleanup of `access_control/test/test_model_validation.py`

### CODE REVIEW

File: swagger_server/__main__.py [hunk 1]
```diff
@@ -31,6 +32,9 @@
 app.app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = settings.SQLALCHEMY_TRACK_MODIFICATIONS
 
 app.add_error_handler(SQLAlchemyError, exception_handlers.db_exceptions)
+app.add_error_handler(
+    controller_validators.InvalidRequest, controller_validators.handle_invalid_request
```
Comment (cobusc) [inline swagger_server/__main__.py]: Do you need add the exception (`controller_validators.InvalidRequest`) as well? Why not just the handler?

File: swagger_server/controllers/controller_validators.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+from access_control.models import (DeletionMethod, Site)
+
+
+# Exceptions and handlers
+class InvalidRequest(Exception):
+    status_code = 400
+
+    def __init__(self, message, status_code=None, payload=None):
+        Exception.__init__(self)
```
Comment (cobusc) [inline swagger_server/controllers/controller_validators.py]: `super().__init__(self)` is preferred, unless there is a specific reason for using the base class directly.

File: swagger_server/controllers/controller_validators.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+# Validators
+class SiteValidator:
+
+    def validate_site_deletion_method_data(self, lookup_id, data):
+        try:
+            deletion_method = DeletionMethod.query.filter_by(
+                id=lookup_id
+            ).one()
+        except NoResultFound:
+            raise InvalidRequest(f"No result found for: DeletionMethod<id={lookup_id}>")
```
Comment (cobusc) [inline swagger_server/controllers/controller_validators.py]: "Deletion method with ID {lookup_id} does not exist." seems more userfriendly

File: swagger_server/__main__.py (hunk 1 above)
Comment (AltusBarry) [reply swagger_server/__main__.py]: `add_error_handler` has no optional kwargs. `def add_error_handler(self, error_code, function):`

File: swagger_server/controllers/controller_validators.py (hunk 2 above)
Comment (AltusBarry) [reply swagger_server/controllers/controller_validators.py]: Exasperation fuelled copy from Flask docs. Will update.

File: swagger_server/controllers/controller_validators.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        try:
+            jsonschema.validate(
+                data,
+                schema=deletion_method.data_schema,
+                format_checker=jsonschema.FormatChecker()
+            )
+        except jsonschema.ValidationError:
+            raise InvalidRequest(
+                f"deletion_method_data validation failed against schema: {deletion_method.data_schema}"
```
Comment (cobusc) [inline swagger_server/controllers/controller_validators.py]: "The deletion method data does not conform to the schema defined for the deletion method {deletion_method.label}"

File: swagger_server/controllers/controller_validators.py (hunk 3 above)
Comment (AltusBarry) [reply swagger_server/controllers/controller_validators.py]: Will update

File: access_control/models.py [hunk 5]
```diff
@@ -154,26 +152,22 @@ def __repr__(self):
...
-                instance = DeletionMethod.query.filter_by(
-                    id=kwargs["deletion_method_id"]
-                ).one()
-            except NoResultFound:
-                abort(
-                    400,
-                    f"No result found for: DeletionMethod<id={self.deletion_method_id}>"
-                )
+    def __validate_deletion_method_data(self, lookup_id, data):
+        # Due to oddities in the orm level validation, this method is called
```
[the remark this answers is not in the document]
Comment (AltusBarry) [reply access_control/models.py]: Thanks

File: swagger_server/controllers/controller_validators.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            )
+
+    def validate_site_deletion_method_required_fields(self, lookup_id, data):
+        has_data = data or data == {}
+        if has_data and not lookup_id >= 0:
```
Comment (cobusc) [inline swagger_server/controllers/controller_validators.py]: `if data is not None and lookup_id is None` is much clearer for me.

File: swagger_server/controllers/controller_validators.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            raise InvalidRequest("deletion_method_id is required if deletion_method_data is provided")
+        elif lookup_id >= 0 and not has_data:
```
Comment (cobusc) [inline swagger_server/controllers/controller_validators.py]: `elif lookup_id is not None and data is None:`

File: swagger_server/controllers/controller_validators.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        )
+        self.validate_site_deletion_method_data(
+            method_id, method_data
+        )
+
+    def validate_site_update(self, lookup_id, data):
+        try:
+            site = Site.query.filter_by(id=lookup_id).one()
+        except NoResultFound:
+            raise InvalidRequest(f"No result found for: Site<id={lookup_id}>")
```
Comment (cobusc) [inline swagger_server/controllers/controller_validators.py]: "Site with ID {lookup_id} does not exist"

File: swagger_server/test/__init__.py [hunk 9]
```diff
@@ -58,6 +59,10 @@ def create_app(self):
         flask_app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
         DB.init_app(flask_app)
         app.add_error_handler(SQLAlchemyError, exception_handlers.db_exceptions)
+        app.add_error_handler(
+            controller_validators.InvalidRequest,
```
Comment (cobusc) [inline swagger_server/test/__init__.py]: Why is it necessary to add the exception here as well as the handler?

Comment (AltusBarry) [reply swagger_server/test/__init__.py]: See previous response to the one in __main__. Connexion add_error_handler has no optional kwargs, it requires both.

File: swagger_server/controllers/controller_validators.py (hunk 4 above)
Comment (AltusBarry) [reply swagger_server/controllers/controller_validators.py]: Will update

File: swagger_server/controllers/controller_validators.py (hunk 8 above)
Comment (AltusBarry) [reply swagger_server/controllers/controller_validators.py]: Will update

File: swagger_server/controllers/controller_validators.py (hunk 6 above)
Comment (AltusBarry) [reply swagger_server/controllers/controller_validators.py]: Will update

File: swagger_server/controllers/controller_validators.py (hunk 7 above)
Comment (AltusBarry) [reply swagger_server/controllers/controller_validators.py]: will update

**Comment by cobusc** [comment]
:+1:

**Pull request closed without merging**